Add support for REAPI execution from the sandbox

juergbi opened the pull request:
Fixes #1985.

juergbi committed:
Enable local execution in buildbox-casd to support nested execution

This bumps the minimum version of buildbox-casd to 1.2.0.

juergbi committed:
sandbox: Add support for `remote-apis-socket`

Fixes #1985.

juergbi committed:
_context.py: Add `effective_build_max_jobs` helper

abderrahim commented:
I tested this, and recc inside the sandbox is complaining
```
2025-05-27T08:22:40.382+0000 [218:140519756722176] [buildboxcommon_grpcretrier.cpp:134] [ERROR] ActionCache.GetActionResult() failed with: 12: Action Cache service is not enabled for instance ""
2025-05-27T08:22:40.383+0000 [218:140519756722176] [executioncontext.cpp:467] [ERROR] Error while querying action cache at "unix:/tmp/casd.sock": 12: Action Cache service is not enabled for instance ""
2025-05-27T08:22:40.385+0000 [218:140519756722176] [buildboxcommon_grpcretrier.cpp:134] [ERROR] Execution.Execute() failed with: 12: Execute: 12: Action Cache service is not enabled for instance ""
2025-05-27T08:22:40.386+0000 [218:140519756722176] [executioncontext.cpp:832] [ERROR] Error while calling `Execute()` on "unix:/tmp/casd.sock": 12: Execute: 12: Action Cache service is not enabled for instance ""
```

I dug a bit and it is due to the fact that I'm using a cache storage service. Disabling it works fine.

gtristan commented on src/buildstream/sandbox/_config.py:
```diff
@@ -66,15 +74,15 @@ def __init__(
         #     breaking cache key stability.
         #
-        sandbox_dict: Dict[str, Union[str, int]] = {"build-os": self.build_os, "build-arch": self.build_arch}
+        sandbox_dict: Dict[str, Union[str, int, bool]] = {"build-os": self.build_os, "build-arch": self.build_arch}
```
Is this outdated perhaps ? Maybe there was an added boolean here in a previous iteration of the patch ?

I think here there is only the possibility of a new string being serialized in the dict.

juergbi replied:
Yes, there was indeed an additional boolean config option in a previous iteration. Will remove this.

gtristan commented:
Generally like this patch overall... although it is still in draft, do we want this in 2.5 ?

Should we add an integration test for exercising `recc` ? Would be pretty cool if we had an example for this in the canonical format too ?

I'd be happy to add a tested example based on the [recc demo](https://gitlab.com/tristanvb/buildstream-recc-demo/) to sweeten the pot if we think we're ready to go ahead with this...

abderrahim commented:
I feel it's still not ready. While it's good enough, ~~it depends on unreleased buildbox~~ (buildbox was released earlier today, my bad) and there are a few corner cases such as https://github.com/apache/buildstream/pull/2014#issuecomment-2911646294 above that need to be handled.

I'd say let's leave it till 2.6. I think that's also @juergbi's opinion as this is still marked as draft.

juergbi commented:
Leaving it till 2.6 is fine with me. I anyway expect typical use cases to require #1991 and that's also still a draft.

juergbi committed:
tests/integration/sandbox.py: Add test for `remote-apis-socket`

juergbi commented:
> I dug a bit and it is due to the fact that I'm using a cache storage service. Disabling it works fine.

I've added a check for this to fail the build with a clear error message. In https://github.com/apache/buildstream/tree/jbilleter/action-cache-service, I'm adding support for `action-cache-service`, which then enables use of the REAPI socket with `storage-service`. That branch is not yet tested (or documented), though, and I think it's ok to first merge this and then the `action-cache-service` addition.

I've also added a minimal test to this branch.

CI is currently failing due to BuildBox being too old for the added test. https://github.com/apache/buildstream/pull/2023 updates the CI images based on https://gitlab.com/BuildStream/buildstream-docker-images/-/merge_requests/230.

abderrahim commented on tests/integration/project/elements/sandbox/remote-apis-socket.bst:
```diff
@@ -0,0 +1,13 @@
+
+config:
+  build-commands:
+    - test -e /tmp/reapi.sock
```
Can't we test that it's a socket instead?

```suggestion
    - test -S /tmp/reapi.sock
```

juergbi replied:
Good suggestion, changed.

abderrahim commented:
> I'm adding support for action-cache-service, which then enables use of the REAPI socket with storage-service. That branch is not yet tested (or documented), though, and I think it's ok to first merge this and then the action-cache-service addition.

Yeah, this is fine. With the clear error message here, we can land this even if it doesn't support the cache storage-service.

I'll test the action-cache-service branch in the mean time

abderrahim commented on src/buildstream/_cas/casdprocessmanager.py:
```diff
@@ -45,9 +45,9 @@
-_REQUIRED_CASD_MINOR = 0
-_REQUIRED_CASD_MICRO = 58
+_REQUIRED_CASD_MAJOR = 1
+_REQUIRED_CASD_MINOR = 2
```
I think we can have some cleanups of fallback code paths now that we require buildbox 1.2. That shouldn't block this pull request, but it should be nice to have it done before the next release.
